Feat: allow users to customize their own rpc urls

Jdecristi opened the pull request:
## General Changes

Fixes #440 

## Developer Notes

I had initially planned to allow users to use their own custom RPC Urls through wallet providers. However, to use WalletConnect, a custom provider is required, which means that the wallet provider alone is not sufficient. To address this issue, I propose creating a modal that allows users to enter an Alchemy or Infura key and customize RPC URLs for each network. This solution provides a secure and flexible way for users to connect their wallets to the application.


## Screen shots

<details>
<summary>Before</summary>
<br />

### Desktop
<img width="1500" alt="Screenshot" src="https://user-images.githubusercontent.com/89173284/235553892-385caa6f-b8cb-4bd4-a34d-608b20ca07a9.png">

---

### Mobile
<img width="400" alt="before-mobile" src="https://user-images.githubusercontent.com/89173284/235554044-41f3752f-e7ce-4ca0-bcbb-273eb98ce672.png">

</details>

<details>
<summary>After</summary>
<br />

### Desktop
<img width="1500" alt="Screenshot 2023-05-05 at 7 24 49 PM" src="https://user-images.githubusercontent.com/89173284/236584583-15f4f6de-562e-481a-8779-41725d80df9a.png">
<img width="1500" alt="Screenshot 2023-05-05 at 7 27 40 PM" src="https://user-images.githubusercontent.com/89173284/236586646-b1512159-365a-497c-98e5-370f15c8da71.png">
<img width="1500" alt="Screenshot 2023-05-11 at 2 43 52 AM" src="https://github.com/aave/interface/assets/89173284/6aa65f5d-adee-4dd3-b394-94ec23eda1c5">
<img width="1500" alt="Screenshot 2023-05-11 at 2 44 21 AM" src="https://github.com/aave/interface/assets/89173284/f68d63a2-2b6d-447c-b25a-da788ef189ab">
<img width="1500" alt="Screenshot 2023-05-05 at 7 58 06 PM" src="https://user-images.githubusercontent.com/89173284/236586558-6b110d3d-c021-4fec-a0af-115504a2c206.png">
.githubusercontent.com/89173284/236584684-589de74b-9a8d-46cc-90a3-db5ddbcb52f4.png">


---

### Mobile
<img width="350" alt="Screenshot 2023-05-05 at 8 00 00 PM" src="https://user-images.githubusercontent.com/89173284/236586917-a718d0e5-45c4-4a3e-893c-df194405b951.png">
<img width="350" alt="Screenshot 2023-05-05 at 8 00 43 PM" src="https://user-ima
<img width="350" alt="Screenshot 2023-05-11 at 2 46 38 AM" src="https://github.com/aave/interface/assets/89173284/ccdd6529-36bb-472c-b312-de5c43bafa01">
<img width="350" alt="Screenshot 2023-05-11 at 2 46 21 AM" src="https://github.com/aave/interface/assets/89173284/21fb3f83-d026-42af-aefb-5cf746d19384">
<img width="350" alt="Screenshot 2023-05-11 at 2 46 38 AM" src="https://github.com/aave/interface/assets/89173284/64130d27-ad08-42a1-ae32-03bce2c6d20c">
<img width="350" alt="Screenshot 2023-05-05 at 8 02 56 PM" src="https://user-images.githubusercontent.com/89173284/236586912-5d7770a0-2061-4fdb-827e-082d2dcf3daa.png">


</details>

---

## Author Checklist

Please ensure you, the author, have gone through this checklist to ensure there is an efficient workflow for the reviewers.

- [x]  The base branch is set to `main`
- [x]  The title is using [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) formatting
- [x]  The Github issue has been linked to the PR in the Development section
- [x]  The General Changes section has been filled out
- [x]  Developer Notes have been added (optional)

**If the PR is ready for review:**

- [x]  The PR is in `Open` state and not in `Draft` mode
- [x]  The `Ready for Dev Review` label has been added

## Reviewer Checklist

Please ensure you, as the reviewer(s), have gone through this checklist to ensure that the code changes are ready to ship safely and to help mitigate any downstream issues that may occur.

- [ ]  End-to-end tests are passing without any errors
- [ ]  Code style generally follows existing patterns
- [ ]  Code changes do not significantly increase the application bundle size
- [ ]  If there are new 3rd-party packages, they do not introduce potential security threats
- [ ]  If there are new environment variables being added, they have been added to the `.env.example` file as well as the pertinant `.github/actions/*` files
- [ ]  There are no CI changes, or they have been approved by the DevOps and Engineering team(s)
- [ ]  Code changes have been quality checked in the ephemeral URL
- [ ]  QA verification has been completed
- [ ]  There are two or more approvals from the core team
- [ ]  Squash and merge has been checked

defispartan commented:
Wow, this looks like an awesome feature!

iamanastasia commented:
Hey @Jdecristi thanks for the pr! A huge problem would be solved with this one.

Few comments:
1. I'm a bit confused with switcher as entry point to URLs setup. Maybe we can add another button there to separate enable-button and setup window? Here is how it would look like for zero state, enabled, disabled.
![Zero state](https://user-images.githubusercontent.com/92739089/235672304-51fe138e-a776-428f-a375-519b306049fd.png)
![Enabled state](https://user-images.githubusercontent.com/92739089/235672617-92d39c23-9ec2-45e6-b96a-408b006e87ed.png)
![Disabled state](https://user-images.githubusercontent.com/92739089/235672633-de50c114-aa37-4651-bd23-b8663961c0e9.png)

2. I recommend removing gradients from the buttons and replace them with primary colors. Also made some changes to text sizes and placeholders to make it more ~~boring~~ clean 🙃.
![Dialog 1](https://user-images.githubusercontent.com/92739089/235672940-ae4a7453-d3c7-4505-9a90-9fe5e0d4a518.png)
![Dialog 2](https://user-images.githubusercontent.com/92739089/235673142-3ba06cd1-3213-42cd-964b-8d08a38bf45f.png)

Let me know what you think!

Jdecristi commented:
@iamanastasia  I have made some changes to your requests and here's what I did:

1. Instead of the Switcher, I replaced it with a Button. I tested various configurations with both buttons and switchers, and the button configuration proved to be the best.

2. I made adjustments to the font sizes to match your preferences and also fixed the placeholders.

3. I removed all the testnets since I realized they were not necessary. Additionally, trying to incorporate the testnets with the mainnets was quite challenging, and their inclusion would not add much value.

Thank you for your feedback, the feature is now much more refined thanks to your input.

### Screenshots

iamanastasia commented:
@Jdecristi Thanks! Button seems to be more obvious. 

Still have concerns, but not sure about use cases:
1. I liked your previous idea with switcher because it would be easy to enable/disable custom RPC when user needs it — imagine user wants to repay debt quickly, but network is clogged, it would take user 1 click to switch the connection.
2. With the button it's not obvious whether I have custom rpc on or off. It says 'Edit', but it doesn't mean it's enabled. Might be confusing, takes some time to learn how it works, some double-checks, adds some friction. 

Let's ask for second opinion? 

@defispartan @MareskoY @0x4Graham Could you please review screenshots fixed on top and screenshots I attached to previous comment? What do you think? Do you think it's a one time setting and setup button is sufficient?

MareskoY commented:
Hello, @Jdecristi

There are a couple of issues that I've noticed with the interface:

1. It doesn't seem to be adapted for dark mode. Here's a screenshot for reference:

<img width="526" alt="Screenshot 2023-05-03 at 12 38 18" src="https://user-images.githubusercontent.com/18032275/235905671-b85f36ae-132b-4f63-8676-e01db33d9ef8.png">

2. There doesn't appear to be any setup for testnet rpc. It would be great to have this option included, and perhaps we could make it so that the testnet inputs are only visible when testnet is enabled, rather than showing everything on one screen.

0x4Graham commented:
I am a big fan of the slider and would prefer that over the button. The rest looks pretty good to me!

Thanks for picking this task up @Jdecristi!

JoaquinBattilana commented:
Awesome feature!
I like the idea of being able to active and deactivate the custom RPCs with a simple click, since sometimes when my RPC fails i need to go and change it and it takes some time, if i wanted to do a quick action i would be doomed. This is something similar to Frame wallet where you can switch the rpc from secondary to primary and its super confortable if one of them is not working correctly.

Jdecristi commented:
Hey Ya'll I fixed the issues you brought up

@MareskoY I fixed DarkMode issues, aswell as testnet support through a toggle.

@JoaquinBattilana @0x4Graham @iamanastasia  I switched it back to a toggle in the settings menu.

--- 

Screenshots are updated above

0x4Graham commented:
Agreed on the button and the switcher. When testing, i found it confusing that I would need to turn the switcher off if I want ed to change the RPC URL

Jdecristi commented:
@iamanastasia @0x4Graham I fixed your request by adding a button as well, and added the feature that when there are no urls present the switcher is hidden.

#### Screenshots above

0x4Graham commented:
Looks great! Thanks @Jdecristi. We'll do the code review as well as QA review!

Jdecristi commented:
Hey @0x4Graham I managed to fix most of the stuff in your comment

1. I added validation to both the URL and the Api Key inputs. The modal will only save when urls/keys are tested and passed.
2. I intended for the user to be able to upload multiple RPC Urls/Api keys but I fixed it so that the URL takes precedence. I think we should keep the feature to allow multiple Urls/Api Keys incase one of their Urls goes down.
3. I fixed it so that Testnet Inputs show up when testnets are enabled.

#### Screenshots at the top

0x4Graham commented:
Hey @Jdecristi Thanks for the updates!

We will take the PR from here to polish it for final shipping.

Thanks for your input and amazing work on this PR!

JoaquinBattilana commented:
Old, not relevant atm.
